File: JS/Medium/BinaryTree/PreOrder/pathSum3.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number}
 */
var pathSum = function (root, targetSum) {
  let count = 0;
  let map = new Map();
  map.set(0, 1); //base case
  //Calculate Prefix Sum & store curr Sum in hash map
  let dfs = (curr, currSum) => {
    if (!curr) return;
    currSum += curr.val;

    //check if there is a valid path
    if (map.get(currSum - targetSum)) {
      count += map.get(currSum - targetSum);
    }

    //add curr sum to map
    map.set(currSum, (map.get(currSum) || 0) + 1);

    dfs(curr.left, currSum);
    dfs(curr.right, currSum);

    //backtrack
    map.set(currSum, map.get(currSum) - 1);
  };
  dfs(root, 0);

  return count;
};

File: C#/DSA/Medium/Trees/pathSum3.cs

//  Definition for a binary tree node.
//   public class TreeNode {
//       public int val;
//       public TreeNode left;
//       public TreeNode right;
//       public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
//           this.val = val;
//           this.left = left;
//           this.right = right;
//       }
//   }

public class Solution3
{
    public int PathSum(TreeNode root, int targetSum)
    {
        int count = 0;
        Dictionary<long, int> map = [];

        map[0] = 1; //base case

        void dfs(long currSum, TreeNode curr)
        {
            if (curr is null) return;
            currSum += curr.val;

            //check if there is valid path
            if (map.TryGetValue(currSum - targetSum, out int found))
            {
                count += found;
            }
            //similar to push
            if (map.TryGetValue(currSum, out int valToIncrease))
            {
                map[currSum] = valToIncrease + 1;
            }
            else
            {
                map[currSum] = 1;
            }

            dfs(currSum, curr.left);
            dfs(currSum, curr.right);


            //backtrack
            //similar to pop
            map[currSum]--;

        }

        dfs(0, root);
        return count;
    }
}